server/audit_math/suite.py

```diff
@@ -0,0 +1,691 @@
+    own contest object, with its own margin. Strata, along with the overall
+    contest object, are passed to the SUITE module when perfoming mixed-strata
+    audits.
+    """
+
+    num_ballots: int
+    vote_totals: Dict[str, int]
+    misstatements: Dict[Tuple[str, str], Dict[str, int]]
+    results: CVRS
+    sample_size: int
+
+    def __init__(
+        self,
+        num_ballots: int,
+        vote_totals: Dict[str, int],
+        results: CVRS,
+        misstatements: Dict[Tuple[str, str], Dict[str, int]],
+        sample_size: int,
+    ):
+        """
+        Initializes the ballot comparison stratum.
+
+        Inputs:
+            contest: contest information for this stratum, including candidate
+                     vote totals and stratum size
+            results: The CVRs for this stratum.
+            misstatements: All of the misstatements observed so far, by winner-loser pair. E.g.,
+                    {
+                        (winner1, loser1): {"o1": 0, "o2": 0, "u1": 0, "u2": 0},
+                        (winner1, loser2): ...,
+                        ...,
+                        (winner2, loser1): ...
+                    }
+            sample_size: the number of ballots sampled so far
+        """
+
+        self.num_ballots = num_ballots
+        self.vote_totals = vote_totals
+        self.results = results
+        self.misstatements = misstatements
+        self.sample_size = sample_size
+
+    def compute_pvalue(self, reported_margin, winner, loser, null_lambda) -> float:
+
+        """
+        Compute a p-value for a winner-loser pair for this strata based on its math type.
+
+        Inputs:
+            reported_margin: the alternative hypothesis margin, in ballots
+            winner, loser: the winner-loser pair to evaluate the hypothesis on
+            null_lambda: the Fisher's combining lambda for the null hypothesis.
+
+        Outputs:
+            pvalue - the pvalue for the hypothesis given the null_lambda
+        """
+
+        if self.sample_size == 0:
+            return 1
+
+        o1, o2, u1, u2 = (
+            self.misstatements[(winner, loser)]["o1"],
+            self.misstatements[(winner, loser)]["o2"],
+            self.misstatements[(winner, loser)]["u1"],
+            self.misstatements[(winner, loser)]["u2"],
+        )
+
+        U_s = 2 * self.num_ballots / reported_margin
+        log_pvalue = (
+            self.sample_size * np.log(1 - null_lambda / (GAMMA * U_s))
+            - o1 * np.log(1 - 1 / (2 * GAMMA))
+            - o2 * np.log(1 - 1 / GAMMA)
+            - u1 * np.log(1 + 1 / (2 * GAMMA))
+            - u2 * np.log(1 + 1 / GAMMA)
+        )
+        pvalue = np.exp(log_pvalue)
+        return float(np.min([pvalue, 1.0]))  # cast for the typechecker
+
+
+def get_misstatements(contest, reported_cvr, sample, winner, loser):
+    o1, o2, u1, u2 = 0, 0, 0, 0
+    for ballot in reported_cvr:
+        if ballot not in sample:
+            continue
```

Comment:
this seems a bit different from the logic in supersimple, where we iterate over the sample results, not the reported results, and also have some special casing for a few different edge cases

Reply:
My thinking about the difference is that in a total ballot comparison audit, there may be ballots that don't have CVRs and aren't included in the results, which may get sampled. Here, it seems like those ballots would get dumped into the ballot polling stratum and so it's safe to assume that there aren't any ballots that don't have CVRs here. I suppose I could just iterate over the sample, I think this was left over from some of my thinking. 

Actually this function is never used, I might just take it out altogether. It should probably move to the api code that initializes these objects.